dm/syncer/binlog_locations.go

```diff
@@ -0,0 +1,178 @@
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	return l.curEndLocation
+}
+
+//nolint:unused
+func (l *locationRecorder) setCurEndLocation(location binlog.Location) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	l.curEndLocation = location
+}
+
+func (l *locationRecorder) saveTxnEndLocation() {
+	l.txnEndLocation = l.curEndLocation
+	_ = l.txnEndLocation.SetGTID(l.curEndLocation.GetGTID().Origin().Clone())
+}
+
+// shouldUpdatePos returns true when the given event is from a real upstream writing, returns false when the event is
+// header, heartbeat, etc.
+func shouldUpdatePos(e *replication.BinlogEvent) bool {
+	switch e.Header.EventType {
+	case replication.FORMAT_DESCRIPTION_EVENT, replication.HEARTBEAT_EVENT, replication.IGNORABLE_EVENT,
+		replication.PREVIOUS_GTIDS_EVENT, replication.MARIADB_GTID_LIST_EVENT:
+		return false
+	}
+	//nolint:gosimple
+	if e.Header.Flags&replication.LOG_EVENT_ARTIFICIAL_F != 0 {
+		// ignore events with LOG_EVENT_ARTIFICIAL_F flag(0x0020) set
+		// ref: https://dev.mysql.com/doc/internals/en/binlog-event-flag.html
+		return false
+	}
+
+	return true
+}
+
+func (l *locationRecorder) setCurrentGTID(gset mysql.GTIDSet) {
+	err := l.curEndLocation.SetGTID(gset)
+	if err != nil {
+		log.L().DPanic("failed to set GTID set",
+			zap.Any("GTID set", gset),
+			zap.Error(err))
+	}
+}
+
+// update maintains the member of locationRecorder as their definitions.
+// - curStartLocation is assigned to curEndLocation
+// - curEndLocation is tried to be updated in-place
+// - txnEndLocation is assigned to curEndLocation when `e` is the last event of a transaction.
+func (l *locationRecorder) update(e *replication.BinlogEvent) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	l.curStartLocation = l.curEndLocation
+
+	if !shouldUpdatePos(e) {
+		// so for these events, we'll set streamer to let curStartLocation equals curEndLocation
+		return
+	}
+
+	if event, ok := e.Event.(*replication.RotateEvent); ok {
+		nextName := string(event.NextLogName)
+		if l.curEndLocation.Position.Name != nextName {
+			l.curEndLocation.Position.Name = nextName
+			l.curEndLocation.Position.Pos = 4
+			l.saveTxnEndLocation()
+		}
+		return
+	}
+
+	l.curEndLocation.Position.Pos = e.Header.LogPos
+
+	switch ev := e.Event.(type) {
+	case *replication.XIDEvent:
+		// for transactional engines like InnoDB, COMMIT is xid event
+		l.setCurrentGTID(ev.GSet)
+		l.saveTxnEndLocation()
+		l.inDML = false
+	case *replication.QueryEvent:
+		query := strings.TrimSpace(string(ev.Query))
+		switch query {
+		case "BEGIN":
+			// MySQL will write a "BEGIN" query event when it starts a DML transaction, we use this event to distinguish
+			// DML query event which comes from a session binlog_format = STATEMENT.
+			// But MariaDB will not write "BEGIN" query event, we simply hope user should not do that.
+			l.inDML = true
+		case "COMMIT":
+			// for non-transactional engines like MyISAM, COMMIT is query event
+			l.inDML = false
+		}
```

Comment:
seems loss the default case for normal ddl query event ?

Reply:
what do you want to do in default branch

Reply:
to update `l.inDML` = false , after think i know that `l.inDML` is already maintained correctly in current implementation

but i think is not so straufgtfirward, it takes a little more thought to understand that we don't need to deal with nore ddl query event hrere

Reply:
I can't get what to explain for default branch. A query event can be DDL or DML so we just keep `inDML` unchanged. the flip of `inDML` is triggered by "BEGIN", "COMMIT" and Xid